io-engine/src/bdev/nexus/nexus_bdev_children.rs

```diff
@@ -739,59 +749,53 @@ fn update_failfast_done(
 }
 
 impl<'n> Nexus<'n> {
-    /// Marks a child device as faulted.
-    /// Returns true if the child was in open state, false otherwise.
-    fn child_io_faulted(
-        self: Pin<&mut Self>,
-        device_name: &str,
-        reason: Reason,
-    ) -> bool {
-        match self.lookup_child_device(device_name) {
-            Some(c) => {
-                debug!("{:?}: faulting with {}...", c, reason);
-
-                if Ok(ChildState::Open)
-                    == c.state.compare_exchange(
-                        ChildState::Open,
-                        ChildState::Faulted(reason),
-                    )
-                {
-                    warn!("{:?}: I/O faulted; will retire", c);
-                    true
-                } else {
-                    warn!("{:?}: I/O faulted; child was already faulted", c);
-                    false
-                }
-            }
-            None => {
-                error!(
-                    "{:?}: trying to fault a child device which \
-                        cannot be not found '{}'",
-                    self, device_name
-                );
-                false
-            }
-        }
-    }
-
-    /// TODO
+    /// Faults the device by its name, with the given fault reason.
+    /// The faulted device is scheduled to be retired.
     pub(crate) fn retire_child_device(
         mut self: Pin<&mut Self>,
         child_device: &str,
-        reason: Reason,
+        reason: FaultReason,
         retry: bool,
     ) {
-        // check if this child needs to be retired
-        let need_retire = self.as_mut().child_io_faulted(child_device, reason);
-
-        // The child state was not faulted yet, so this is the first I/O
-        // to this child for which we encountered an error.
-        if need_retire {
-            Reactors::master().send_future(Nexus::child_retire_routine(
-                self.name.clone(),
-                child_device.to_owned(),
-                retry,
-            ));
+        let name = self.name.clone();
+
+        if let Some(c) = self.as_mut().lookup_child_device_mut(child_device) {
+            if Ok(ChildState::Open)
+                == c.state.compare_exchange(
+                    ChildState::Open,
+                    ChildState::Faulted(reason),
+                )
+            {
+                warn!("{:?}: faulted with {}, will retire", c, reason);
+
+                if let Some(log) = c.start_rebuild_log() {
```

Comment:
Is there a race condition here? If we have >1 concurrent retires, the first retire "wins" and sets state to faulted. But before we attach the rebuild log the other retires would return. Could there be a situation where we have newer IO's before we attach the log?

Reply:
Just to clarify, so you mean there's no chance newer IO's will happen before we attach the log?

Reply:
Sounds like a possible race. If I understand correctly:

```
  cpu0	                           cpu1
  ———	                           ———

child_retire(Child1)
                                     child_retire(Child1)
Success = set_state_faulted
                                     is_already_faulted - return
start_rebuild_log()
attach_log_to_iochannel_of_child1()
                                      new_write_op(W)
                                      attach_log_to_iochannel_of_child1()
```

The rebuild_log will miss logging `W`.

Reply:
I added a comment about this situation on the most recent version of my code.

(W) won't really miss. There two possibilities:
1. (W) goes okay. Data goes to the device, nothing missed, no need to log actually.
2. (W) fails. In this case, It goes down the retire path again but on the different channel, and gets the log (very same log created before), and logs the op.

Also, any failed write I/O is resubmitted if any child fails to write, until all children confirm the write. Failed children are removed in the process, so a write I/O succeeds (or all children are gone).
Every resubmission means the same op will be logged again.

In the most recent change, I eliminated two races I had before:
1) create a new /get an existing log on a child
2) Logging an op on a repeated retire.